### ISSUE TITLE
Problems with full screen under X

### ISSUE THREAD

**Issue by bolle732** [stsp/dosemu2#394]
This is on Ubuntu 16.04 64 bit with devel.
I get two problems when running DOSEmu2 with X in full screen, which works fine with SDL.

1. If a program uses VESA modes with more the 256 colours, switching to full screen results in a black screen, returning to windowed mode restores fine

2. Switching to full screen with 2 monitors covers both monitors, but the resulting stretching of the content makes it unusable

Regarding the first problem, I haven't found a real hint to now. Could be a problem regarding true colour windows.

For the second problem, I had a look at the source in "plugin/X/X.c". The function DisplayWidth() returns the sum of the width of both monitors (2x 1920px = 3840px). Even WidthOfScreen() and there like gives the same value of both together. SDL seems to recognise both monitors independent and covers only the first monitor.

It seems that it's tricky to get X full screen right. I've found an example using "_NET_WM_STATE_FULLSCREEN" which works like SDL doing full screen only on the first monitor.

As I see from the code, Xlib is used. Is it still worth to debug this code further ? As I read, Xcb should be the new way. Maybe with Cairo ? What are you thinking on this ? Or should be the real near future pure SDL ?

Remark: I rarely use full screen. It's not a show stopper for me.

**Comment by stsp** [comment]
Please see here:
https://github.com/stsp/dosemu2/pull/8

**Comment by stsp** [comment]
As for vesa bug, please provide the test-case.

**Comment by bolle732** [comment]
Oh, I've overlooked the Xrandr when I did a search in the issues. Could be a good start for the fullscreen on a single monitor. Should there be a configurable option to set the monitor ?

The full screen problem with high colour VESA modes can be reproduced with "vtest.exe" (see  #301). Use for example mode 160h or 161h.

**Comment by stsp** [comment]
> Should there be a configurable option to set the monitor ?

Probably not, why not just to use the current one?

**Comment by stsp** [comment]
You should look into an xrandr2 branch
and see what code it has there. I really
am not aware about its current state.

**Comment by stsp** [comment]
All I get from vtest.exe is
`Current configuration does not support ReturnSvgaMode.`

**Comment by stsp** [comment]
Well, obviously the last mode it lists
for me is `15eh`. IIRC you are adding some
custom vesa modes.

**Comment by stsp** [comment]
In which case this should be qualified as notabug?

**Comment by bolle732** [comment]
Ok, I can reproduce it with 14ah without custom modes specified. Unfortunately, vtest.exe does not support 8bit colours...
Will see, if there are other test program for DOS around.

**Comment by bolle732** [comment]
I've tested with vbetest.exe from SciTech and it looks the same. Up to 8 bit, full screen is ok. 15 and more bit, full screen is black / blank.

**Comment by stsp** [comment]
Allright.
Would you mind writing a couple of 16toXX_all
remappers? :)

**Comment by stsp** [comment]
Done some preparations to make your
work on this easier. :)

**Comment by bolle732** [comment]
Sorry for my late answer. Haven't been able to get a look at the weekend.
Give me a little time to study a little bit more the code of the remapper.

**Comment by bolle732** [comment]
So, I've finally found some time for a first try of a 16-bit high colour mapper to 32-bit true colour. I know now, why the re-mappers are not beloved. I had some lock-ups where I needed to reboot my machine. DOSEmu2 crashed, but the full screen window stayed with keyboard and mouse grab on...

The code below works on my machine, but I'm not really happy with it, as there are to many variables used due casting the pointers.
```
/*
 * 16 bit true color --> 32 bit true color
 * supports arbitrary scaling
 */
void gen_16to32_all(RemapObject *ro)
{
  int d_x_len;
  int s_x, d_x, d_y;
  int d_scan_len = ro->dst_scan_len;
  int *bre_x;
  int *bre_y = ro->bre_y;

  const unsigned char *src, *src0;
  unsigned char *dst;
  unsigned short *src_2;
  unsigned *dst_4;

  src0 = ro->src_image + ro->src_start;
  dst = ro->dst_image + ro->dst_start + ro->dst_offset;
  d_x_len = ro->dst_width;

  for(d_y = ro->dst_y0; d_y < ro->dst_y1; dst += d_scan_len) {
    src = src0 + bre_y[d_y++];
    src_2 = (unsigned short *) src;
    dst_4 = (unsigned *) dst;
    for(s_x = d_x = 0, bre_x = ro->bre_x; d_x < d_x_len; ) {
      dst_4[d_x++] = bgr_2int(ro->dst_color_space, 5, 6, 5, src_2[s_x]);
      s_x += *(bre_x++);
    }
  }
}
```
@stsp Can you please comment on this ?

Note between: I think, there is room for improvement in case of a crash while in full screen. And it would be a good thing, to not enable full screen if no mapper (xxx_all) is available.

**Comment by stsp** [comment]
Looks good on a first glance!
Note that these remappers are rarely used
as by default we use SDL. So if you don't
know how to make fewer variables and casts,
you can just leave it as is. :)

No, there is no such problem with full-screen.
Missing remapper doesn't mean lock-up, and
you still can use Ctrl-Alt-PgDn to shut down
dosemu. If even this doesn't help, then you
switch to another console and do `killall dosemu.bin`.

**Comment by bolle732** [comment]
Ok, will prepare a pull request tomorrow or so for 15 and 16 bit to 32 bit mapping. Other target colour depths maybe later when I find time.

Regarding the freeze is not of a missing re-mapper. Instead it occurs if I have a re-mapper with wrong code. This occures during development from time to time... DOSEmu2 terminates with stack trace. I didn't tried Ctrl-Alt-PgDn, but Ctrl-Alt-<Arrow> or Ctrl-Alt-F1 to F7 didn't work. The keyboard grab is really strong :(

**Comment by bolle732** [comment]
Regarding the freeze. It is not due a missing re-mapper. It occurred when I had a re-mapper with wrong code. This occurs during development from time to time... DOSEmu2 terminates with stack trace. I didn't tried Ctrl-Alt-PgDn, but Ctrl-Alt-<arrow> or Ctrl-Alt-F1 to F7 didn't work. The keyboard grab was really strong :( At the end, I had to reboot hard my machine. Maybe I need to enable Magic SysRq or something like that.

**Comment by stsp** [comment]
I am not sure why you copy/pasted the
comment to which I already replied.
No, you don't need sysrq for that.

**Comment by bolle732** [comment]
Oh, sorry, seems to be due a caching problem of my Firefox browser / Github site or something similar (the write text box had hold the old entry). The duplicated comment can be deleted, should I?

**Comment by stsp** [comment]
Do you think we still need some multi-head
solution for X, will it give us something over
what SDL offers?

**Comment by bolle732** [comment]
I still prefer X because it is still more lightweight. Less CPU usage means more battery lifetime.

With a multi-head set-up, spanning the full-screen over all monitors is mostly not that use-full. In my case, I use different programs in full-screen mode on a specific monitor of a specific workspace (RDP of local Windows VM, browser with important Website). If this could be possible with DOSEmu2, that would be great. But in this case, it shouldn't grab the keyboard and mouse too...

Regarding this case, I think we should create a list with the most likely scenarios / configurations. If you think this is useful, I can start that on the Wiki here. Even I'm a little short on time ;-)

**Comment by stsp** [comment]
> I still prefer X because it is still more lightweight. Less CPU usage

Is this still the case?
I recently applied a few patches and got
better doom2 performance with sdl than
with X.

> With a multi-head set-up, spanning the full-screen over all monitors is mostly not that use-full.

You said it is not the case with sdl.
So for now sdl suits you very well except
for the performance, is this correct?

> If this could be possible with DOSEmu2, that would be great.

I am confused, IIRC you said it is already possible
with dosemu2 with sdl?

> I think we should create a list with the most likely scenarios / configurations.

What exactly do you mean? Scenarios of the
full-screen use or what?

**Comment by bolle732** [comment]
> Is this still the case?
I recently applied a few patches and got
better doom2 performance with sdl than
with X.

A quick test right now with only moving mouse around:
X ~ 20% CPU, SDL ~30% CPU.
So yes, they are closer then before.

> You said it is not the case with sdl.
So for now sdl suits you very well except
for the performance, is this correct?

Yes for full-screen case I can use SDL.

> I am confused, IIRC you said it is already possible
with dosemu2 with sdl?

Yes for SDL, but still the grab should not be activated in full-screen on a multi head set-up when using only one monitor.

> What exactly do you mean? Scenarios of the
full-screen use or what?

See here ;-) [Brain dump](https://github.com/stsp/dosemu2/wiki/Idea---wish-list-of-windowed---full-screen-modes)

**Comment by stsp** [comment]
> So yes, they are closer then before.

Can you also see with disabled swrend?

> Yes for SDL, but still the grab should not be activated

This is not sdl-specific and is unfixable, right?

> See here ;-)

But I don't understand, is this to point people
what backend they should use and whether
they should use fixed aspect? But I think
everyone already knows his use-case.
Or what exactly does this document?

**Comment by bolle732** [comment]
> Can you also see with disabled swrend?

My test was with "$_SDL_swrend = (off)". Using $_SDL_swrend = (on), the speed is about the half and it doesn't change the CPU usage.

> This is not sdl-specific and is unfixable, right?

Don't know, if it's impossible to disable the grab in full-screen.

How does it Firefox when switching to full-screen with F11 ? There it covers all the actual screen and I can enter leave the mouse without grab and easily switch work spaces. Most video player do the same. Could it be, that this is not a real full-screen mode, only a window resized to the full extend of the current monitor and instructing the window manager to not draw any borders ?

> But I don't understand, is this to point people
what backend they should use and whether
they should use fixed aspect? But I think
everyone already knows his use-case.
Or what exactly does this document?

Should be something like a fact sheet, which could help in development of what configuration options may be needed to accommodate all the cases.

Should I move the wiki page to my fork ? Maybe it's better there, as it still WiP.

**Comment by stsp** [comment]
> Don't know, if it's impossible to disable the grab in full-screen.

Ah, this is of course possible.
You can easily find that code in both
sdl.c and X.c, just search for "grab".
But I wonder how will that differ from
just maximizing the window?
If there are sufficient differences, then
perhaps it worth a new option, so feel
free to add it. :)

> Should be something like a fact sheet, which could help
> in development of what configuration options may be needed
> to accommodate all the cases.

How about removing cases 1, 2 and 5?
Then I think we are very good at covering
all cases. :)

> Should I move the wiki page to my fork ?

Since I don't think anyone is reading it,
I won't really care either way. I just don't
think this can be of any help for the users,
but maybe it will help us to better understand
each other's use cases.

**Comment by stsp** [comment]
Perhaps instead of cases 1 and 2 we
can add a hotkey to reset the window
size to the nearest integral aspect.

As for grab, can't you just release it
with the appropriate hot-key? But of
course a new small option for that
will not hurt.

**Comment by bolle732** [comment]
> But I wonder how will that differ from
just maximizing the window?

Depending on the desktop environment, a maximized window still doesn't cover the start menu / panels and there like.

> How about removing cases 1, 2 and 5?
Then I think we are very good at covering
all cases. :)

For 1) I like the case as it prevents me from accidentally resizing the window. This is maybe the mode I use it mostly... This is where the gen_NNtoNN**_1** remappers under X will be used, where performance / CPU usage would be optimized.

For 2) This is like the already existing mode of zooming the content by the factor of 2. Only that it should be extended to allow an arbitrary fixed zoom. Should be useful especially for HiDPI displays which will become more common in the future. And no accidentally resizing.

For 5) I would use this where the DOS graphic resolution is only a little bit smaller then the real display resolution. In this case, a small black border doesn't hurt. Imagine for example running a DOS CAD program with 1280x1024 on a 1920x1080 monitor. Such an application is sensitive to the aspect ratio / WYSIWYG.

> Perhaps instead of cases 1 and 2 we
can add a hotkey to reset the window
size to the nearest integral aspect.

I really would prefer to not be able to resize the window.

> As for grab, can't you just release it
with the appropriate hot-key? But of
course a new small option for that
will not hurt.

Under X it doesn't ungrab the keyboard and mouse, under SDL, it works.

Regarding full-screen / grab, I discovered a special problem under Ubuntu with Unity:

 - Run DOSEmu2 with -X on a laptop and close the lid
 - Open the lid again

One would expect the login screen to appear. Unfortunately, the DOSEmu2 full-screen is still in the front of the lock screen and can be accessed. In my case, I have mounted / lredired my documents for the program running under DOSEmu2... I haven't checked if this problem exist under SDL too.

Problems like the above are the reason, why I try to investigate what different options are available. As I personally currently still prefer X, this include xcb, cairo, randr, glx etc. This would maybe one of the next brain dump I try. To make decisions or bring good ideas, I need some overview / big pictures and that needs time.

**Comment by stsp** [comment]
> For 2) This is like the already existing mode of

... of X, not of SDL.
The fundamental rule, as I understand, should
be that we do not control the window size. You
can't portably make it to resize by the steps you
want, even if X allows that. So the best we can
do is to use the black borders, in which case
1 and 5 are valid and basically are the same
thing, but 2 should IMO be considered invalid
case unless you can easily code it up with SDL.
So if we take the fact that window always resizes
but we can prevent resizing an image (use
borders), then the full-screen handling is the
same as windowed and you can shrink your
table twice. And to cover all your problems
we need just 2 new options: one that disables
the grab mode under full-screen, and another
one that disables scaling.

> Under X it doesn't ungrab the keyboard and mouse, under SDL, it works.

Then its a bug, you can report it and/or fix it. :)

> To make decisions or bring good ideas, I need some overview / big pictures and that needs time.

Please make sure the good ideas do not
require adding too many new options/code.

**Comment by bolle732** [comment]
Thanks a lot for improving the fullscreen mode.

Unfortunately, I've the problem now, that I can't move the mouse horizontally. Vertically works, but it keeps at x=0...

**Comment by bolle732** [comment]
More tests with "-X":

 - Leaving the keyboard grab allows to switch work spaces
 - Leaving the mouse grab allows the mouse to leave the full-screen window, but then, it refuses to enter the X window again...

SDL seems to be OK.

**Comment by bolle732** [comment]
> Unfortunately, I've the problem now, that I can't move the mouse horizontally. Vertically works, but it keeps at x=0...

On my laptop at home, mouse is working correctly in full-screen with X. The difference to the other laptop is, that it's single head, not multi head.

**Comment by stsp** [comment]
Yes, so I can't reproduce this...
Would you mind fixing it yourself please?

**Comment by bolle732** [comment]
We'll see what I can find. Or maybe I better donate a second monitor to you :-)

**Comment by bolle732** [comment]
A quick fix to get the mouse working in full-screen under X on my 2 head set-up with 2 identical monitors:
```
diff --git a/src/plugin/X/X.c b/src/plugin/X/X.c
index 73c474d..18eb64d 100644
--- a/src/plugin/X/X.c
+++ b/src/plugin/X/X.c
@@ -2269,7 +2269,7 @@ static void X_vidmode(int w, int h, int *new_width, int *new_height)
 {
   int nw, nh, mx, my, shift_x, shift_y;
 
-  nw = DisplayWidth(display, screen);
+  nw = DisplayWidth(display, screen) / 2;
   nh = DisplayHeight(display, screen);
 
 #ifdef HAVE_XVIDMODE
```
Seems like a problem with real width (2 monitors) and X full-screen window (1 monitor). Is this where Xrandr would be needed ?

**Comment by stsp** [comment]
I don't think SDL uses xrandr, or does it?
I don't know what your fix does, unfortunately.

**Comment by bolle732** [comment]
> I don't think SDL uses xrandr, or does it?

Should not be used by default, but maybe it could:
https://wiki.libsdl.org/SDL_HINT_VIDEO_X11_XRANDR

> I don't know what your fix does, unfortunately.

It's not a fix (I've changed the wording in my post), but a workaround for my set-up with 2 monitors.
It could be seen as a hint from where the problem probably comes. I think, we get at one place the screen width for 1 monitor and at another for 2 monitors and this 2 different X values are somehow used in the mouse code.

I've intended to check this today early, but real live hit me badly... I start to get an aversion for MCE on server :(
